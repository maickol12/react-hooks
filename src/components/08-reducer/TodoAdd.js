import React from 'react'
import { useForm } from '../../hooks/useForm';

export const TodoAdd = ({ handleAddTodo}) => {
    const [values, handleInputChange, reset] = useForm({ description: '' });

    const handleSubmit = (e) =>{
        e.preventDefault();
        
        if(values.description.trim().length <= 1){
            return;
        }

        const newTodo = {
            id: new Date().getTime(),
            desc: values.description,
            done: false
        };
    

        handleAddTodo(newTodo);
        reset();
    }
    return (
        <>
           <h4>Agregar todo</h4>
            <hr />
            <form onSubmit={handleSubmit}>
                <input
                    className='form-control'
                    type='text'
                    value={ values.description }
                    onChange={ handleInputChange }
                    name='description'
                    placeholder='Aprender...'
                    autoComplete='off'></input>
                <button type='submit' className='btn btn-primary mt-1 btn-block'>
                    Agregar
                </button>
            </form>  
        </>
    )
}
