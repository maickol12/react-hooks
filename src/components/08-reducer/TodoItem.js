import React from 'react'

export const TodoItem = ({value,handleToggle,deleteTodo,indice}) => {
    const { id, done, desc } = value;
    return (
        <li key={id} className='list-group-item'>
            <p 
                className={`${done && 'complete'}`}
                onClick={() => handleToggle(id) }>
                {indice + 1}. {desc}
            </p>
            <button 
                onClick={ () => deleteTodo(id) }
                className='btn btn-danger'>
                Borrar
            </button>
        </li>
        
    )
}
