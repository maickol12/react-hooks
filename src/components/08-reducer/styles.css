body{
    padding: 70px;
}

p{
    margin: 0px;
    cursor: pointer;
}

.complete{
    text-decoration: line-through;
}

.list-group-item{
    align-items: center;
    display: flex;
    justify-content: space-between;
}