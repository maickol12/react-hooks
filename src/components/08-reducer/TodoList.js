import React from 'react';
import { TodoItem } from './TodoItem';

export const TodoList = ({todos,handleToggle,deleteTodo}) => {
    return (
        <div className='col-5'>
            <ul className='list-group list-group-flush'>
            { todos.map((value,indice) => (
                       <TodoItem
                            key={`todoItem${value.id}`}
                            value={value}
                            handleToggle={handleToggle}
                            deleteTodo={deleteTodo}
                            indice={indice} /> 
                    )
                    ) 
                }
            </ul>
        </div>
    )
}
