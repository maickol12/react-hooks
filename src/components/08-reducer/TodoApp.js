import React, { useReducer, useState, useEffect } from 'react'
import { useForm } from '../../hooks/useForm';
import './styles.css';
import { TodoAdd } from './TodoAdd';
import { TodoList } from './TodoList';
import { todoReducer } from './todoReducer';


const init = () => {
    return JSON.parse(localStorage.getItem('todos')) || []
    // return [{
    //     id: new Date().getTime(),
    //     desc: 'Aprendiendo react',
    //     done: false
    // }];
}


export const TodoApp = () => {

    const [ todos, dispatch ] = useReducer(todoReducer, [], init);

    useEffect(() => {
        localStorage.setItem('todos',JSON.stringify(todos));
    }, [todos])



    const deleteTodo = (id) => {
        
        const action = {
            type:'delete',
            payload: id
        }

        dispatch(action);

    }

    const handleToggle = (id) => {
        dispatch({type:'toggle',payload:id});
    }
    
    const handleAddTodo = (newTodo) => {
        const action = {
            type:'add',
            payload: newTodo
        }
        dispatch(action);
    }

    return (
        <div>
            <h1>TodoAPP ({ todos.length})</h1>
            <hr />
            <div className='row'>
                <TodoList 
                    todos={todos}
                    handleToggle={handleToggle}
                    deleteTodo={deleteTodo} />
                <div className='col-7'>
                   <TodoAdd handleAddTodo={handleAddTodo} />
                </div>

            </div>
        </div>
    )
}
