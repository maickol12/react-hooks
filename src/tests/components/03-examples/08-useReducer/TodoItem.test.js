import { shallow } from "enzyme"
import { TodoItem } from "../../../../components/08-reducer/TodoItem"

const data = {id:1, done:true, desc:'Realizando pruebas'};

describe('Pruebas en <TodoItem />', () => {
    const handleToggle = jest.fn();
    const handleDelete = jest.fn();
    const wrapper = shallow(
        <TodoItem value={data} handleToggle={handleToggle} deleteTodo={handleDelete} indice={0} />
    );

    test('debe de mostrarse correctamente', () => {
        expect( wrapper ).toMatchSnapshot();
    });

    test('debe de llamar la funcion borrar ', () => {
        wrapper.find('button').simulate('click');
        expect( handleDelete ).toHaveBeenCalledWith( 1 );
    });

    test('debe de llamar la funcion toggle ', () => {
        wrapper.find('p').simulate('click');
        expect( handleToggle ).toHaveBeenCalledWith( 1 );
    });

    test('debe de mostrar el texto correcto', () => {
        const p = wrapper.find('p');
        expect( p.text() ).toBe('1. Realizando pruebas')
    });

    test('debe de tener la clase complete si el todo.done es true', () => {
        expect(wrapper.find('p').hasClass('complete')).toBe(true);
    });
    
    
})
