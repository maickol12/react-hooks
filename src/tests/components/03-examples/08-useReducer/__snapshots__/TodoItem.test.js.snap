// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <TodoItem /> debe de mostrarse correctamente 1`] = `
<li
  className="list-group-item"
  key="1"
>
  <p
    className="complete"
    onClick={[Function]}
  >
    1
    . 
    Realizando pruebas
  </p>
  <button
    className="btn btn-danger"
    onClick={[Function]}
  >
    Borrar
  </button>
</li>
`;
